curvefs/src/client/s3/client_s3_cache_manager.cpp

```diff
@@ -0,0 +1,1092 @@
+/*
+ *  Copyright (c) 2020 NetEase Inc.
+ *
+ *  Licensed under the Apache License, Version 2.0 (the "License");
+ *  you may not use this file except in compliance with the License.
+ *  You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ */
+
+/*
+ * Project: curve
+ * Created Date: 21-8-18
+ * Author: huyao
+ */
+
+#include "curvefs/src/client/s3/client_s3_cache_manager.h"
+#include "curvefs/src/client/s3/client_s3_adaptor.h"
+
+namespace curvefs {
+namespace client {
+
+std::string GenerateObjectName(uint64_t chunkId, uint64_t blockIndex) {
+    std::ostringstream oss;
+    oss << chunkId << "_" << blockIndex << "_0";
+    return oss.str();
+}
+
+FileCacheManagerPtr FsCacheManager::FindFileCacheManager(uint64_t inodeId) {
+    ReadLockGuard readLockGuard(rwLock_);
+
+    auto it = fileCacheManagerMap_.find(inodeId);
+    if (it != fileCacheManagerMap_.end()) {
+        return it->second;
+    }
+
+    return NULL;
+}
+FileCacheManagerPtr FsCacheManager::CreateFileCacheManager(uint64_t fsId,
+                                                           uint64_t inodeId) {
+    FileCacheManagerPtr fileCacheManager =
+        std::make_shared<FileCacheManager>(fsId, inodeId, s3ClientAdaptor_);
+
+    WriteLockGuard writeLockGuard(rwLock_);
+    fileCacheManagerMap_.emplace(inodeId, fileCacheManager);
+
+    return fileCacheManager;
+}
+
+void FsCacheManager::ReleaseFileCacheManager(uint64_t inodeId) {
+    WriteLockGuard writeLockGuard(rwLock_);
+
+    auto iter = fileCacheManagerMap_.find(inodeId);
+    assert(iter != fileCacheManagerMap_.end());
+
+    fileCacheManagerMap_.erase(iter);
+    return;
+}
+
+std::list<DataCachePtr>::iterator FsCacheManager::Set(DataCachePtr dataCache) {
+    WriteLockGuard writeLockGuard(rwLockLru_);
+
+    if (lruReadDataCacheList_.size() == lruCapacity_) {
+        dataCache = lruReadDataCacheList_.back();
+        lruReadDataCacheList_.pop_back();
+        dataCache->Release();
+    }
+    lruReadDataCacheList_.push_front(dataCache);
+    return lruReadDataCacheList_.begin();
+}
+
+void FsCacheManager::Get(std::list<DataCachePtr>::iterator iter) {
+    WriteLockGuard writeLockGuard(rwLockLru_);
+
+    lruReadDataCacheList_.splice(lruReadDataCacheList_.begin(),
```

Comment:
Spilce source and destination are the same, thus maybe cause a problem, do we test it when lruReadDataCacheList_ only one element?